### PR TITLE
Make scan to be a cython extention

### PR DESCRIPTION
**Pull request by ferrine**
**Thank you for opening a PR!**
Address #78 to make scan be a Cython extension

* To install properly with pip I had to use pyproject.toml which is a go-to recommendation. With setup.cfg I struggled to specify build time dependencies

## Before
1. User installs Cython and a compiler at install time
2. At first import time compiles Scan

## After
1. We build Scan at bdist_wheel time
2. We create a bdist_wheel matrix on the github actions as done in e.g. [scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/main/.github/workflows/wheels.yml#L50)/[pandas](https://github.com/pandas-dev/pandas/blob/main/.github/workflows/wheels.yml#L46)
3. User installs the compiled code or from sdist (compiles on his own during pip install), explained [here](https://medium.com/ochrona/understanding-python-package-distribution-types-25d53308a9a)

Here are a few important guidelines and requirements to check before your PR can be merged:
+ [ ] There is an informative high-level description of the changes.
+ [ ] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [ ] [`pre-commit`](https://pre-commit.com/#installation) is installed and [set up](https://pre-commit.com/#3-install-the-git-hook-scripts).
+ [ ] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [ ] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), and there are **no commits that fix changes introduced by other commits in the same branch/BR**.
+ [ ] There are tests covering the changes introduced in the PR.

Don't worry, your PR doesn't need to be in perfect order to submit it.  As development progresses and/or reviewers request changes, you can always [rewrite the history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) of your feature/PR branches.

If your PR is an ongoing effort and you would like to involve us in the process, simply make it a [draft PR](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests).

### CODE REVIEW

**Comment by ricardoV94** [comment]
Will this be doing compilation all the time or is it cached?

**Comment by ferrine** [comment]
> Will this be doing compilation all the time or is it cached?

It will compile once on installation and it requires 
1) install time Cython dependency. We can omit scan_perform.c from git
2) cythonize scan_perform and declare Extension via setuptools, no Cython Dependency

I like the Cython first approach more, less maintainance + a lot of libraries require Cython anyway. 

The Cython install tyime dependecy can be stated in pyproject.toml so user never takes care of installing Cython before he installs pytensor

**Comment by twiecki** [comment]
With hddm we supported both. If Cython was installed, we would cythonize to c, if not, the shipped c-extensions would be used: https://github.com/hddm-devs/hddm/blob/master/setup.py#L4

**Comment by ferrine** [comment]
> With hddm we supported both. If Cython was installed, we would cythonize to c, if not, the shipped c-extensions would be used: https://github.com/hddm-devs/hddm/blob/master/setup.py#L4

pyproject.toml allows to specify buld time dependencies so if you pip install without having some packages like numpy or Cython it will install them for you to run setup.py in an isolated mode. After installation they will not appear in the userspace. So you can only have Cython to compile the extension and be always sure it will be available

**Comment by aseyboldt** [comment]
Did you maybe forget to git add `pytensor/scan/scan_perform.pyx`?
I don't think I get it yet. Why do we want to compile some part of the scan code at package build time? We don't know what the scan ops are supposed to do at that time, right?

**Comment by ferrine** [comment]
@aseyboldt the scan Op as it is written in the package accepts storage for outputs and a void function f()
It iteratively calls f(), checks outputs and returns. It does not seem to have any interaction outside its scope

**Comment by ferrine** [comment]
> Did you maybe forget to git add `pytensor/scan/scan_perform.pyx`?

It was already there, I think I better delete some generated C code instead and reorganize imports quite a bit

File: .github/workflows/pypi.yml [hunk 1]
```diff
@@ -18,27 +18,43 @@ concurrency:
 
 jobs:
   build:
+    strategy:
+      matrix:
+        python-version: ["3.7", "3.8", "3.9", "3.10"]
```
Comment (ricardoV94) [inline .github/workflows/pypi.yml:23]: I think we were only testing earliest and latest release. Adding 2 extra ones may be too much woth our very long tests.

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    "Topic :: Software Development :: Compilers",
+    "Topic :: Scientific/Engineering :: Mathematics",
+    "Operating System :: Microsoft :: Windows",
+    "Operating System :: POSIX",
+    "Operating System :: Unix",
+    "Operating System :: MacOS",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.7",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
```
Comment (ricardoV94) [inline pyproject.toml:39]: Missing 3.10 here

File: pytensor/_version.py [hunk 3]
```diff
@@ -105,7 +102,7 @@ def run_command(commands, args, cwd=None, verbose=False, hide_stderr=False, env=
             return None, None
     else:
         if verbose:
-            print(f"unable to find command, tried {commands}")
+            print("unable to find command, tried %s" % (commands,))
```
Comment (ricardoV94) [inline pytensor/_version.py:105]: Why this?

**Comment by ricardoV94** [comment]
@ferrine can we get a clean git history for this one? Seems too important to just merge squash everything.

File: .github/workflows/pypi.yml (hunk 1 above)
Comment (ferrine) [reply .github/workflows/pypi.yml:23]: At this point we can delegate the compilation to user or precompile on our side:

What is included in the matrix is precompiled, the reset is compiled from the source dist.

The most flexible and realistic use-case for us is to delegate everything to user and supply pypi only with sdist, not bdist

Comment (ferrine) [reply .github/workflows/pypi.yml:23]: Anyway the user has to compile some C code or scan at some point

Comment (ferrine) [reply .github/workflows/pypi.yml:23]: > I think we were only testing earliest and latest release. Adding 2 extra ones may be too much worth our very long tests.

This only tests you can import pytensor + builds linux wheels

File: pytensor/_version.py (hunk 3 above)
Comment (ferrine) [reply pytensor/_version.py:105]: It was a versioneer script refresh, I've pinned the version in the pyproject.toml

**Comment by ferrine** [comment]
@ricardoV94 I've cleaned the history, and removed the code on that we are not going to depend:

* The C generated code using Cython (Cythin is installed by setuptools)
* The wrapper over `scan_perform.c` that checked if the extension is already compiled

**Comment by ferrine** [comment]
The last major question I think is important is:

"Who compiles scan now given it was compiled by user before?"

Available options
* Users compile (old behaviour)
* Users on Linux + python 3.7-3.10(11) do not compile, Windows/MacOS users do compile
* Users on [Linux, Windows, MacOS(x86), MacOS(arm)] x [py3.7, py3.8, py3.9, py3.10, py3.11] do not compile (whatever we manage to provide)

**Comment by aseyboldt** [comment]
> @aseyboldt the scan Op as it is written in the package accepts storage for outputs and a void function f()
It iteratively calls f(), checks outputs and returns. It does not seem to have any interaction outside its scope

I still don't get it yet. Previously this was compiled at runtime, right? So the c compiler would see the function in the loop and could optimize the loop with that in mind. But if this is precompiled, it can no longer specialize at all. That doesn't sound like what we want to me, or am I missing something?

**Comment by ricardoV94** [comment]
> I still don't get it yet. Previously this was compiled at runtime, right? So the c compiler would see the function in the loop and could optimize the loop with that in mind. But if this is precompiled, it can no longer specialize at all. That doesn't sound like what we want to me, or am I missing something?

My understanding is that the general Scan Cython code is frozen. Only the inner function is compiled at runtime. I'll let @ferrine confirm that though.

**Comment by ferrine** [comment]
> What's the benefit from this PR? We save on shipping the C code? Anything else?

1. We remove legacy code to support on demand compilation for scan which was a bit hackish
2. We provide an example to write Cython extensions for future developers
3. We provide all the requirements to add a custom Cython extension if any is needed to extend numba
4. https://github.com/pymc-devs/pytensor/pull/46/files#diff-18626e6b1322e1b5dda375d2401417044a4e25d94041ca894059f1b2db878f14R140 gives the path to custom Cython extensions that interact with numba

I think this PR will make the code base slightly more structured which we would want it to be in the long term. Even if we remove Scan C code in the future there is a setup to support Cython extensions. If we prefer scan over generated numba code, we wrap it, if we prefer to drop Scan it is also fine.

**Comment by aseyboldt** [comment]
I'm still somewhat hung up on precompilation before distribution. This complicates distribution, because pytensor is no longer a pure python package then. We have to build the extension module for different architectures and operating systems, and can't easily use more recent machine code instructions. Why would we want to do this in the first place?
I looked through the old source for scan, and that clearly doesn't seem ideal, where something might be compiled at import time, but I don't think moving that into the package creation time is helpful. I would really like to get rid of this, instead of move it to a different place.
I'd really love to get this a place where pytensor is just a pure python package, and we can move all the pain of having to deal with distribution of compiled libraries to numba or other backend packages. This seems like a step away from this, because it introduces distribution / compilation pain that we didn't have before. From the PR:

```
  # The job to build pypi wheels users can installed precompiled for them.
  # At the moment only linux wheels are build, Windows and MacOS will compile on installation
  # Before prebuilds all users had to compile code anyway.
```
This is true currently, but we want to get rid of that "users have to compile code anyway".

Building a few Cython extensions is more managable than all the other compiler shenanigans we have to deal with, so if there really is a long-term need for our own extension modules, I think we can manage, but I really don't see the need.

**Comment by ferrine** [comment]
@ricardoV94 @aseyboldt Will Scan code be removed at some point?

**Comment by ricardoV94** [comment]
> @ricardoV94 @aseyboldt Will Scan code be removed at some point?

Only when we remove the C backend. I would imagine it will stay for a minimum of one year more.

**Comment by aseyboldt** [comment]
Been thinking it over a bit..

> Only when we remove the C backend. I would imagine it will stay for a minimum of one year more.

I think you are right, and if so I think this cython extension is probably a good think after all. I guess I've been a bit too optimistic about how fast we can remove the c backend...

So sorry @ferrine, I think you are right :-)

**Comment by ferrine** [comment]
Got merge conflicts resolved

**Comment by ferrine** [comment]
Can we merge this?

**Review by ricardoV94 — APPROVED** [review]

File: scripts/run_mypy.py [hunk 4]
```diff
@@ -61,6 +61,9 @@
 pytensor/tensor/type.py
 pytensor/tensor/type_other.py
 pytensor/tensor/var.py
+"""
+# fixed a some point
+"""
 pytensor/typed_list/basic.py
```
Comment (ricardoV94) [inline scripts/run_mypy.py:67]: Just remove it?

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

**Comment by ricardoV94** [comment]
Your first commit title seems to have a typo? "refresgithub actions"

And the last two commits should go together I think.

Otherwise I'll trust you because I am not familiar with these internals. Pinging @maresb in case he can spot anything.

**Review by ricardoV94 — APPROVED** [review]

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    "pytensor/assert_op.py",
+    "pytensor/graph/opt.py",
+    "pytensor/graph/opt_utils.py",
+    "pytensor/graph/optdb.py",
+    "pytensor/graph/kanren.py",
+    "pytensor/graph/unify.py",
+    "pytensor/link/jax/jax_linker.py",
+    "pytensor/link/jax/jax_dispatch.py",
+    "pytensor/graph/toolbox.py",
+    "pytensor/scalar/basic_scipy.py",
```
Comment (michaelosthege) [inline pyproject.toml:152]: Some of these get deleted in your other PR, right?

We should remember to update this after rebasing the other PR

**Review by michaelosthege — APPROVED** [review]
I suppose you copied the changes to `versioneer.py` and we don't need to review these?

**Comment by maresb** [comment]
Sorry, I'm just seeing this now. What did you want me to have a look at? Something in the `run_mypy.py` script? Or did you solve it? (I see that CI is green now.)

**Comment by michaelosthege** [comment]
> Sorry, I'm just seeing this now. What did you want me to have a look at? Something in the `run_mypy.py` script? Or did you solve it? (I see that CI is green now.)

If you could check the `pyproject.toml` and changes in GitHub actions that'd be great

**Comment by maresb** [comment]
`DESCRIPTION.txt` is RST, so probably better to rename the file so that it can be parsed properly.

**Review by maresb — APPROVED** [review]

File: pyproject.toml (hunk 5 above)
Comment (ferrine) [reply pyproject.toml:152]: yeah, one of these PRs goes first

**Comment by maresb** [comment]
:point_up: fixed in https://github.com/pymc-devs/pytensor/pull/134

**Comment by ferrine** [comment]
I did rebase an merge since there were some major blocks of chages

**Comment by ricardoV94** [comment]
> I did rebase an merge since there were some major blocks of changes

That should be the action in most of the PRs ;)

**Pull request merged**